data/db/clickhouse.py

```diff
@@ -0,0 +1,184 @@
+            )
+
+        if Clickhouse.__schema is None:
+            Clickhouse.__schema = Clickhouse.load_schema()
+
+    def load_schema():
+        raw_schema = yaml.safe_load(open(SCHEMA_PATH, 'r'))
+
+        clickhouse_schema = {}
+        relationship_to_table = {}
+
+        for s in raw_schema:
+            table_name = raw_schema[s]['db_collection_name']
+            properties = raw_schema[s]['properties']
+            relationship = raw_schema[s].get('relationship')
+            relationship_to_table[s] = table_name
+
+            if table_name in clickhouse_schema:
+                clickhouse_schema[table_name]['properties'].update(properties)
+            else:
+                clickhouse_schema[table_name] = {'properties': properties}
+
+            # Example of a relationship config block from schema-config.yaml:
+            # relationship:
+            #   from: genes
+            #   to: genes, mm_genes
+            # must be converted into the following columns for Clickhouse:
+            # from: ['genes_1_id'], to: ['genes_2_id', 'mm_genes_id']
+            if relationship:
+                clickhouse_schema[table_name]['relationship'] = {
+                    'from': [i.strip() for i in relationship['from'].split(',')],
+                    'to': [i.strip() for i in relationship['to'].split(',')]
+                }
+
+                froms = set(
+                    clickhouse_schema[table_name]['relationship']['from'])
+                tos = set(clickhouse_schema[table_name]['relationship']['to'])
+
+                clickhouse_schema[table_name]['relationship']['from'] = [
+                    f + '_id' if (f not in tos) else f + '_1_id' for f in froms]
+                clickhouse_schema[table_name]['relationship']['to'] = [
+                    f + '_id' if (f not in froms) else f + '_2_id' for f in tos]
+
+        return clickhouse_schema
+
+    def get_connection():
+        return Clickhouse.__connection
+
+    def get_schema():
+        return Clickhouse.__schema
+
+    def process_json_line(self, json_line, properties):
+        data = json.loads(json_line)
+        processed_data = []
+
+        for key in properties:
+            # biocypher heritage for number cases
+            if key not in data:
+                key += ':long'
+
+            # certain properties might not exist in all records, setting them to None
+            value = data.get(key)
+
+            # clickhouse discourages the use of NULL, using '' instead
+            if value == None or value == []:
+                value = ''
+            processed_data.append(value)
+
+        processed_data.append(data['_key'])
+        if data.get('_from') is not None and data.get('_to') is not None:
+            # removing collection prefix from _from and _to. E.g. genes/ENSG00000148584 => ENSG00000148584
+            processed_data.append(data['_from'].split('/')[-1])
+            processed_data.append(data['_to'].split('/')[-1])
```

Comment:
how do you figure out which collection to go if there are more than one collection in from or to?

Reply:
at this time, we don't need the combinatorial aspect, we just need to convert that list of tables into ID columns for Clickhouse. For example:

from: genes, mm_genes
to: transcripts, mm_transcripts

must be converted into:

genes_id String
mm_genes_id String
transcripts_id String
mm_transcripts_id String

as foreign keys in the SQL schema.

Reply:
good catch, will look into this!